liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Axiom.hs

```diff
@@ -32,6 +33,24 @@ import           Language.Haskell.Liquid.Types
 
 import           Language.Haskell.Liquid.Bare.Resolve as Bare
 import           Language.Haskell.Liquid.Bare.Types   as Bare
+import qualified Data.Set as Set
+import Language.Haskell.Liquid.Misc (fst4)
+import Control.Applicative
+
+-- Generalized function to find duplicates
+findDuplicate :: (Ord a) => Set.Set a -> [a] -> Maybe (a, a)
+findDuplicate _ [] = Nothing
+findDuplicate seen (x:xs)
+  | Just y <- Set.lookupLE x seen, y == x = Just (y, x)
+  | otherwise                             = findDuplicate (Set.insert x seen) xs
+
+-- Find duplicates within a single list
+findDuplicatePair :: (Ord a) => [a] -> Maybe (a, a)
+findDuplicatePair = findDuplicate Set.empty
```

Comment:
Perhaps this is simpler:
```suggestion
findDuplicatePair :: Eq a => [a] -> Maybe (a, a)
findDuplicatePair xs = listToMaybe [ (a, b)  | a:b:_ <- group (sort xs) ]
```

I don't understand why the element is output twice. Is there a need to yield both components if they are equal?

Reply:
Yes because essentially they can be deemed equal w/o being definitionally equal, which will be the case here. The equality will be performed on the symbol names but not on their location. Since we want to get the location of both "equal" elements if we have duplicates, we return the two duplicate elements so that we can get their location.

Reply:
> Yes because essentially they can be deemed equal w/o being definitionally equal, which will be the case here. The equality will be performed on the symbol names but not on their location. Since we want to get the location of both "equal" elements if we have duplicates, we return the two duplicate elements so that we can get their location.

More precisely because of the way equality is defined on located elements:
```
instance Eq a => Eq (Located a) where
  (Loc _ _ x) == (Loc _ _ y) = x == y
```